docs/architecture.md

```diff
@@ -0,0 +1,147 @@
+
+It enables the creation of Kata Containers VMs on any machines without the need for bare metal worker nodes,
+or nested virtualisation support.
+
+## Background
+
+With the confidential-containers projects we are aiming to deliver a confidential computing enabled Kubernetes Platform
+providing:
+- **Security and Confidentiality** - We want to provide highly secure solutions to our users' applications, limiting the
+actions of the cloud service provider (CSP), or Kubernetes Admins and providing assurance that their data and workloads
+remain confidential and safe from modification.
+- **Performance** - We want to maximise performance, whilst accepting that there is a trade-off to gain extra security.
+- **Compatibility with established Cloud Native Applications** - We don't want our users to have to re-write their
+workloads just to get confidential computing features, but to be able to bring along any Kubernetes workload with
+minimal changes.
+
+The way that we do this is utilise Trusted Execution Environments (TEEs). There are areas of hardware where code is
+executed with the following features:
+- Hardware root-of-trust
+- Remote Attestation
+- Trusted launch
+- Memory encrypted and isolated
+- Secure key management
+
+Many TEEs such as SEV, TDX and Secure Execution use virtual machines as the trusted boundary of the hardware, and in
+order to be compatible with existing Kubernetes deployments, we put the TEE boundary around the Pod Sandbox, not the
+whole Kubernetes node, so Kubernetes' orchestration can still work without modification, so we have some different
+options of how to lay out this infrastructure.
+
+![Kubernetes Infrastructure Models](./KubernetesInfrastructureModels.png)
+
+### Bare Metal
+
+In order make the TEE VM a first-level KVM guest, the current way to set-up Kubernetes infrastructure is to have a
+whole bare metal machine as the the Kubernetes node and inside of that we can have 1 or more TEE VMs, in which we
+can run our pod sandboxes.
+Although this is the traditional way to currently configure a Kata Containers VMs in the cloud, there is a big
+drawback of having to have a separate bare metal box per Kubernetes node, which is expensive and difficult to scale.
+
+
+### Nested Virtualisation
+
+Most current cloud providers need to have the scalabilty and cost savings that bare metal nodes cannot provide, so
+utilise virtual servers. These partition up the physical processing power and resources into multiple systems that
+can each run their own operating system. A common model in non-confidential Kubernetes infrastructure is to have a
+Kubernetes node map to a virtual server, so each physical server can run multiple nodes and each of them could run
+multiple pod sandboxes. Currently there are limitation of not being able to run as a second-level VM in a performant
+way for most TEEs and more CSP, means we can't put these pod sandboxes inside a TEE boundary in these scenarios.
```

Comment:
I didn't understand this "and more CSP". But it's Friday and perhaps my brain is already in weekend mode.

Reply:
Sorry, that should be most CSPs (as Azure have added nested support)